Feature/cubic interpolation 359

lelemmen opened the pull request.

kimt33 commented on flik/linesearch/interpolations.py:
```diff
@@ -0,0 +1,109 @@
+def cubic_interpolation(var, func, grad, direction, interval_start, interval_end, alpha_i_, alpha_i):
+    """
+    A cubic interpolation algorithm for functions where the gradient is not expensive to calculate, according to eq.
+    (3.59) in "Numerical Optimization" by Nocedal and Wright.
```
The first line in the docstring should not extend beyond one line. You can move the reference to the References section in the docstring.

kimt33 commented on flik/linesearch/interpolations.py:
```diff
@@ -0,0 +1,109 @@
+
+    Parameters
+    ----------
+    var: np.array((N,1), dtype=float)
```
It's not a two dimensional array
`np.ndarray((N,), dtype=float)`

kimt33 commented on flik/linesearch/interpolations.py:
```diff
@@ -0,0 +1,109 @@
+
+
+def cubic_interpolation(var, func, grad, direction, interval_start, interval_end, alpha_i_, alpha_i):
+    """
```
Should be a raw string b/c of the latex stuff

kimt33 commented on flik/linesearch/interpolations.py:
```diff
@@ -0,0 +1,109 @@
+            if interval_start and interval_end are not such that
+                0 < interval_start < interval_end <= 1
+            if alpha_1 and alpha_2 are not such that
+                interval_start < alpha_i_ < alpha_i < interval_end
```
I think these all can be indented once to the left

kimt33 commented on flik/linesearch/interpolations.py:
```diff
@@ -0,0 +1,109 @@
+            according to the function check_input
+        ValueError
+            according to the function check_input
+            if interval_start and interval_end are not such that
```
I think you can just say
`If 0 < interval_start < interval_end <= 1`

kimt33 commented on flik/linesearch/interpolations.py:
```diff
@@ -0,0 +1,109 @@
+    # Check the input arguments
+    # TODO: Check the parameters according to the function check_input
+    if isinstance(var, float):
+        var = np.array(var)
```
`np.array(var, dtype=float, ndmin=1)`

kimt33 commented on flik/linesearch/interpolations.py:
```diff
@@ -0,0 +1,109 @@
+        The starting point of the interval containing desirable step lengths
+    interval_end: float
+        The ending point of the interval containing desirable step lengths
+    alpha_i_: float
```
I'm not too sure about the name of this varaible

kimt33 commented on flik/linesearch/interpolations.py:
```diff
@@ -0,0 +1,109 @@
+    print(alpha_i_, phi(alpha_i_), phi_(alpha_i_))
+    print(alpha_i, phi(alpha_i), phi_(alpha_i))
+    print(interval_start, phi(interval_start))
+    print(interval_end, phi(interval_end))
```
I think we can delete all of the print statements (we'll do a proper logging later on I think)

kimt33 commented on flik/linesearch/interpolations.py:
```diff
@@ -0,0 +1,109 @@
+        raise ValueError("alpha_i_ and alpha_i should be according interval_start < alpha_i_ < alpha_i < interval_end")
+
+    # Define the helper functions phi and its derivative
+    def phi(alpha):
```
docstring

kimt33 commented on flik/linesearch/test/test_interpolations.py:
```diff
@@ -0,0 +1,93 @@
+import numpy as np
+from nose.tools import assert_raises
+
+
```
one too many newlines

kimt33 commented on flik/linesearch/test/test_interpolations.py:
```diff
@@ -0,0 +1,93 @@
+
+def test_cubic_interpolation():
+    """Test the cubic interpolation algorithm with a manual calculation."""
+
```
one too many newline

kimt33 commented on flik/linesearch/test/test_interpolations.py:
```diff
@@ -0,0 +1,93 @@
+
+    def grad(x):
+        return 2 * x
+
```
one too many newline (I think)

lelemmen replied:
pycodestyle doesn't complain about this one, though.

kimt33 commented on flik/linesearch/test/test_interpolations.py:
```diff
@@ -0,0 +1,93 @@
+    # alpha_1 = 0.3
+    # alpha_2 = 0.9
+    # assert_raises(ValueError, cubic_interpolation, var, func, grad, direction, interval_start, interval_end, alpha_1,
+    #               alpha_2)
```
do these tests fail?

kimt33 reviewed (changes requested):
Just some small changes. I think Travis will get most of these when you push again (you need to submit it again to get noticed by Travis).

lelemmen commented on flik/linesearch/interpolations.py:
```diff
@@ -93,19 +93,19 @@ def phi_(alpha):
+    alpha_next = alpha_current - (alpha_current - alpha_previous) * (phi_(alpha_current) + d_2 -
+                                                                     d_1) / (phi_(alpha_current) -
+                                                                             phi_(alpha_previous) +
+                                                                             2 * d_2)
```
I really don't agree on enforcing "E127 continuation line over-indented for visual indent" on this one. The formulas are completely illegible.

lelemmen commented:
At this point Travis only fails because of coverage in flik/linesearch/input.py.

Commit:
Add cubic interpolation

1. Add cubic interpolation
2. Add tests

kimt33 committed:
Clean up cubic interpolation

1. Change API
2. Update check_input
3. Add more flexibility for alphas (can be equal to interval)

kimt33 committed:
Allow alpha to be zero

Because certain interpolation algorithms start from alpha=0
